Allow to import GIF files

Daw11 opened the pull request:
Fixes: https://github.com/godotengine/godot-proposals/issues/1433

![godot-gif](https://user-images.githubusercontent.com/3595817/64067219-61e89700-cc25-11e9-8fb9-d2e6a9f9c63e.gif)

This pull allows to directly import GIF files inside Godot.

It adds a module that uses giflib to convert each frame of the GIF to RGBA images.

Inside the third party folder there are only the files of giflib that are required to do this operation ( almost all of the functions are inside dgif_lib.c ). I didn't change the source code of the library, so there are a couple of extra files that are there only because they are required to compile the code.

Right now it only allows to read GIFs, I don't know if saving changes is something that we should handle or not.

I'm not sure of which data structure is the best for this kind of files.
AnimatedTexture is simple and it allows to set a custom delay for each frame of the animation. The problem is that it has a limit of 256 frames.

SpriteFrames has the opposite problem, it doesn't have a limit to the number of frames but it can't set a specific delay for each frame.

Right now they are converted into AnimatedTextures.

Another problem is that there's no way to change the flags of the Texture. To change them you have to manually set them for each frame.

Once completed it will resolve #8599

Calinou commented:
This will also help in solving point 3 in #31683 (GIFs don't display in the asset library).

Daw11 commented:
It seems that right now it compiles only on Linux. I'll try later to fix the other platforms.

Xrayez commented:
> Right now it only allows to read GIFs, I don't know if saving changes is something that we should handle or not.

From the [giflib  README](https://sourceforge.net/p/giflib/code/ci/master/tree/README#l10):

> GIF is a legacy format; we recommend against generating new images in
it.  For a cleaner, more extensible design with better color support
and compression, look up PNG.

Or rather, APNG.

asheraryam commented:
I suggest to convert it to AnimatedSprite if it's less than 256 frames, and if it's more than 256 frames create a Sprite with an attached AnimationPlayer that contains an animation with the delays.

volzhs commented:
> I suggest to convert it to AnimatedSprite if it's less than 256 frames, and if it's more than 256 frames create a Sprite with an attached AnimationPlayer that contains an animation with the delays.

I'd like to suggest to have one way either AnimatedSprite, AnimatedTexture or AnimationPlayer, not have different behavior depending on something.

asheraryam commented:
In that case I'd suggest using Spriteframes + Animationplayer, similar to what the mesh importer does for 3D models.

Xrayez commented:
How easy would it be to load `AnimatedTexture` dynamically this way? Basically the question reflects this https://github.com/godotengine/godot/issues/8599#issuecomment-298264104.

All other image loaders inherit from `ImageFormatLoader` and not `ResourceFormatLoader`, but I see how this is problematic with `Image` class only supporting static images (one can't possibly return something along the lines of `AnimatedImage` though I think it would be bad idea anyway).

At the very least, it would be nice to `image.load('foobar.gif')` loading only the first frame of the GIF.

Daw11 commented:
Loading quickly only one frame is not simple. The data of the GIF right now is loaded with the function `DGifSlurp` which reads all the frames ( they call it high-level API). To read only one at a time and to stop after the first one the function must be replaced with the low-level code of giflib that parse the data of the GIF.
I saw that when a file is converted 50% of the time is taken by giflib and the other half is the creation of the AnimatedTexture, so even if it creates only one frame it will not be a lot faster.
I'll try to see if I can replace the function with low-level code that parse each frame one by one.

Daw11 commented:
I updated the code and now frames are loaded one by one.
This reduces the memory usage and it will allow to quickly load the first image even on a very big GIF.

Daw11 commented:
I'm not sure if it's the best solution to the problem but to make the importing of GIF files more flexible I'm thinking of adding a new helper class to do the job.

The default behavior of the editor will be the same as right now, when a GIF is dragged  into the scene or is loaded with `load("res://...")` it will be converted into an `AnimatedTexture` without flags.

The new class might be named `AnimatedGIF` and it will contain only an array of Images and an array of frame delay.

It will have the following methods:

```
load_from_file(String path, int max_frames)
load_from_buffer(PoolByteArray data, int max_frames)
```
These can load any number of frames from any GIF.

```
to_animated_texture(int flags, int max_frames)
to_sprite_frames(int flags, int max_frames)
to_animated_sprite_2d(int flags, int max_frames)
to_animated_sprite_3d(int flags, int max_frames)
```
These will generate a new Resource from the loaded `AnimatedGIF`.

```
open(PoolByteArray data);
load_frames(PoolByteArray data, int max_frames);
close()
```
These will probably be added in a another pull, their purpose is to allow the user to dynamically read the frames of a GIF while it's being downloaded, without having to wait for all the data to be ready.

Is this a good solution or there's a better one?

Xrayez commented:
Makes perfect sense for me.

I mean, there's multiple ways you can import a `png` for instance:

![png_multiple_ways](https://user-images.githubusercontent.com/17108460/64633937-d3ce9680-d404-11e9-9e0d-8bf91536ef53.png)

(btw, this shows multiple entries, bug found! which only proves my point nobody is aware of this feature).

The same way GIF could be imported by specifying import option perhaps, and this could be done by creating `EditorImportPlugin` I think.

Also see #31927 for barebones.

Daw11 commented:
I added the new class that allows to load and convert GIFs, right now it's called `AnimatedGIF`.

GIF files can now be imported as `AnimatedTexture` or as `SpriteFrames`. 

![gif-import](https://user-images.githubusercontent.com/3595817/65372922-a12f6400-dc76-11e9-8fca-d350b7a15219.png)

In the import panel the params that you can change are the maximum number of frames and the flags of the texture.
When a GIF is an `AnimatedTexture` all the frames have the correct speed but their maximum number is always 256. With the average framerate of a GIF being 10 FPS this is usually enough.

`AnimatedSprites` don't have that limit but the speed of the animation is always the average of all the frames. It's probably better to allow `SpriteFrames` to have an individual frame delay than to add a workaround right now and import them with an `AnimationPlayer`.

You can still create an animation player with a script that reads the data inside a loaded `AnimatedGIF` class and then save it with `ResourceSaver`.

I think that now this pull is ready.

Xrayez commented:
Tested importing `AnimatedTexture`, works flawlessly! I'd merge.

The only concern is with `AnimatedGIF` name, perhaps it could be generalized to `AnimatedImage`, unless it has some specific code related to GIF format.

Daw11 commented:
@Xrayez I'll change the class name into AnimatedImage and I'll allow it to use different loaders. The only format it can load right now is GIF, another one could be APNG.
Right now the images are imported without flags but I'm wondering if `Mipmaps` and `Filter` should be enabled by default.

Xrayez commented:
Indexed GIFs would likely be used for pixel-art style assets due to limited number of colors, where `Filter` is disabled, as well as `Mipmaps` I guess.

On the other hand, APNG would allow for high-quality animated images.

So likely both `Filter` and `Mipmaps` could be disabled for now.

Daw11 commented:
I completed all the changes.
The class now is called AnimatedImage and it can be used to load multiple formats. The only one it support right now is GIF but I'll try to add APNG too in a different pull.
If a GIF is imported as a SpriteFrames, now it can be dragged and dropped just like AnimatedTexture.

Daw11 committed:
Allow to import GIF files

Use giflib to import GIF images and convert them into AnimatedTexture or Spritefames.

Calinou commented:
PS: If you're continuing work on this, make sure to add `.gif` as a recognized extension in the dummy image importer as I did in https://github.com/godotengine/godot/pull/36083.

asheraryam commented:
Would be nice to have this updated and merged! I want to make a plugin to record GIFs in godot (so people can share them directly) and this class would help a lot.

Xrayez commented:
Note that it's not possible to *save* GIFs with the current implementation, only the loader is implemented.

There's been enough of core changes which introduced merge conflicts which must be resolved now (mostly `Texture2D` refactor, and some concepts as texture filtering are no longer part of the texture as filtering can be set per node in 4.0), but in its current state, it's possible to merge this to 3.2 in order to avoid the pain of backporting this from 4.0 to 3.2.

For that matter, and depending on the approval of this PR from core devs, I'd like @Daw11 to consider contributing to [`ImageTools` module](https://github.com/Xrayez/godot-imagetools) as it currently targets Godot 3.2, or allow myself or other contributors to use your work and adding you as a co-author at the very least. I believe it won't be too much work integrating it with another module because the current implementation *is* a module. I can even add you as a collaborator for such a big feature if you wish. 🙂

asheraryam commented:
@Xrayez Alright thanks for the headsup! I'll probably do this using a C# library instead, that's my go-to now for general utility libraries.

aaronfranke commented:
@Daw11 This looks neat, but it needs to be rebased on the latest master branch for it to be reviewed. After rebasing, you will also have to run `misc/scripts/fix_style.sh --clang-format` to fix style issues, since master has new style rules, including no empty lines after opening braces.

Xrayez commented:
OK, so I'm not sure if the PR is abandoned by now (likely yes), but I'd like to implement this feature in some way or another (while still having this feature useful for my own use cases, of course).

My main motivation for having this is simply being able to `load` GIF via script. I'm not really that interested in various ways on how animated images could be imported, but I think it makes sense to restrict this to `AnimatedTexture` in any case.

So, I created godotengine/godot-proposals#1433 (with an implementation already done as a self-contained module) and even godotengine/godot-proposals#1448 because the implementation in the first proposal was considered a bloat for core developers (mainly the idea of implementing an additional class to expose via script, because I don't really see another way). Any other implementation requires me to do double work: one for my own (specific?) use cases, and one for Godot, and I'd really like to avoid that. We need a community effort to justify the addition of another class dedicated to animated images, or allow to rewrite `Image` class to support animated image data (there are numerous image formats for this).

But well, not implementing this feature for core is also OK for me, as I said I managed to make it fully self-contained C++ module and I already do use a bunch of C++ modules for other stuff already.

Xrayez commented:
Implemented in goostengine/goost#8.

aaronfranke commented:
@Daw11 Is this still desired? If so, it needs to be rebased on the latest master branch. Also, the above feedback should be addressed.

Otherwise, abandoned pull requests will be closed in the future as announced [here](https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable).

reduz commented:
Merging things into core must be done only if absolutely necessary, and backed with a very well defined use case.
As cool as it may seem, It is not the case for GIF which is pretty useless, so I advocate for just closing this. If you ultimately need it, it should be doable with a gdnative plugin.

aaronfranke commented:
@reduz What about the use case of viewing GIF files in the Godot Asset Library? Wouldn't that make this useful for core? It could be made editor-only by default unless someone compiles a custom export template.

Xrayez commented:
> @reduz What about the use case of viewing GIF files in the Godot Asset Library? Wouldn't that make this useful for core? It could be made editor-only by default unless someone compiles a custom export template.

In that sense, yes, GIF support will be useful. Just being able to load or import GIFs as *static* images could also be fine I guess. It would be probably more useful than BMP importer currently implemented in Godot, not sure.

But this likely won't happen because providing a library to import GIFs would ask for even more feature requests, so I guess reduz would like to avoid that...

gitsomebit commented:
what is the current workaround for gif files loaded at runtime? (got user request to use animated gifs for customisation and emotes)
